### DISCUSSION TITLE
Agents without LLM usage

### DISCUSSION THREAD

**Discussion by kaikreuzer** [eclipse-lmos/lmos-demo#18] [General]
I would like to implement a kind of "proxy agent", i.e. one that simply forwards the incoming question to an external chatbot and then returns the received answer back as a response.
I assume Arc is not the right framework to implement this, so it's rather a "bring your own agent" (BYOA) type of problem. Are there already any best practices on how I could build such an agent for LMOS?

**Comment by RobWin** [comment]
The current LMOS Runtime communicates with Arc agents using **GraphQL WebSockets**, but there are plans to transition to a **Web of Things (WoT)** communication protocol. This transition will offer greater flexibility by enabling protocol bindings like HTTP, MQTT, AMQP, or WebSockets. We have two options:

### **Option 1: Implementing the Proxy with the WoT Approach**

1. **Use the WoT Library**:  
   The upcoming **Web of Things Kotlin library** will allow you to implement a proxy agent that provides a WoT-compliant interface for the external chatbot.

2. **WoT Description**:  
   The proxy will expose a **WoT description**, detailing the proxied agent’s capabilities. The LMOS runtime can interact with this proxy using a WoT client.

3. **Protocol Bindings**:  
   - The library supports multiple bindings (e.g., HTTP, MQTT).
   - While the **HTTP binding** is currently functional, work is ongoing for the **MQTT binding**. You could start a proof of concept (PoC) using HTTP and later migrate to MQTT or another protocol as needed.

4. **Advantages of WoT**:  
   - Standardized and interoperable communication.
   - Flexibility in switching protocols based on requirements.

---

### **Option 2: Using the Current GraphQL Approach**

1. **Spring Boot Application**:  
   Develop a **Spring Boot application** that acts as a **GraphQL-to-HTTP proxy**. This application would:
   - Receive GraphQL queries from the LMOS runtime.
   - Translate these queries into HTTP requests for the external chatbot.
   - Process the responses from the chatbot and return them in GraphQL format.

2. **GraphQL Spring Boot Starter**:  
   Use the **GraphQL Spring Boot Starter** to simplify the implementation of the GraphQL API for the LMOS runtime.

3. **Advantages of GraphQL Approach**:  
   - Direct compatibility with the current LMOS setup.
   - No need to modify the runtime immediately.

---

### **Recommendation**
- If you're looking for a quick and temporary solution, go with **Option 2** (GraphQL proxy).
- If you prefer a future-proof and standards-compliant solution, start a **PoC with the WoT library** using the HTTP binding, and adapt the LMOS runtime to use a WoT client.
